File: Projects/top-tier-customs/services/auth/src/middleware/auth.middleware.js
import jwt from "jsonwebtoken";
import ENV from "../config/env.js";

const authorize = (req, res, next) => {
  const header = req.headers["authorization"];
  if (!header) {
    return res.status(401).json({
      source: "<api.auth.middleware>: authorize()",
      message: "Missing Authorization header.",
      success: false,
    });
  }

  const [type, token] = header.split(" ");
  if (type !== "Bearer" || !token) {
    return res.status(400).json({
      source: "<api.auth.middleware>: authorize()",
      message: "Invalid Authorization header format.",
      success: false,
    });
  }

  try {
    const payload = jwt.verify(token, ENV.JWT.ACCESS);

    req.user = payload;

    next();
  } catch (e) {
    return res.status(401).json({
      source: "<api.auth.middleware>: authorize()",
      message: "Invalid or expired Access Token.",
      success: false,
    });
  }
};

export default authorize;

File: Projects/top-tier-customs/services/auth/internal/middleware/auth_middleware.go
package middleware

import (
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/samidmunir/top-tier-customs/services/auth/internal/config"
	"github.com/samidmunir/top-tier-customs/services/auth/internal/utils"
)

func Authorize() gin.HandlerFunc {
	return func(c *gin.Context) {
		authHeader := c.GetHeader("Authorization");
		if authHeader == "" {
			c.JSON(http.StatusUnauthorized, gin.H{
				"ok": false,
				"source": "<api.auth.middleware>: Authorize()",
				"message": "Missing Authorization header.",
			});
			c.Abort();
			return;
		}

		parts := strings.SplitN(authHeader, " ", 2);
		if len(parts) != 2 || strings.ToLower(parts[0]) != "bearer" {
			c.JSON(http.StatusUnauthorized, gin.H{
				"ok": false,
				"source": "<api.auth.middleware>: Authorize()",
				"message": "Invalid Authorization header format.",
			});
			c.Abort();
			return;
		}

		env := config.LoadEnv();

		claims, err := utils.VerifyToken(parts[1], env.JWTAccessSecret);
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{
				"ok": false,
				"source": "<api.auth.middleware>: Authorize()",
				"message": "Invalid or expired access token.",
			});
			c.Abort();
			return;
		}

		c.Set("userId", claims.Sub);
		c.Set("email", claims.Email);
		c.Set("role", claims.Role);

		c.Next();
	}
}